### PR TITLE
rollback: rollback command don't rollback state if block store has one more block

### PR DESCRIPTION
**Pull request by JayT106**
Reopen PR #8101, revised the rollback test case by resetting the privval state. Therefore, the consensus can proceed new block after restarting the node.

### CODE REVIEW

**Comment by JayT106** [comment]
fixing the e2e/app/state.

**Comment by cmwaters** [comment]
Thanks for reopening - sorry that this got lost. I'll bring this up in our meeting today to see what the view is on this

File: test/e2e/app/state.go [hunk 1]
```diff
@@ -7,14 +7,15 @@ import (
...
 	"os"
 	"path/filepath"
 	"sort"
+	"strings"
 	"sync"
 )
 
-const stateFileName = "app_state.json"
-const prevStateFileName = "prev_app_state.json"
+const stateFileName = "app_state_%06d.json"
```
[the remark this answers is not in the document]
Comment (JayT106) [reply test/e2e/app/state.go:18]: just think the e2e tests will not go over 6 digit blocks, doing this mainly for loading the files from the folder, so the files will be sorted perfactly.

**Comment by JayT106** [comment]
@cmwaters are there any concerns from the team about this PR? Haven't heard any feedback.

**Comment by cmwaters** [comment]
It's not so much a concern as I fail to see how rolling back the last block could be useful the moment you restart, the node is just going to fetch that same block again. The only thing that needs to happen is to rerun the last block against the application again. Feel free to correct me if I've missed something.

**Comment by JayT106** [comment]
> It's not so much a concern as I fail to see how rolling back the last block could be useful the moment you restart, the node is just going to fetch that same block again. The only thing that needs to happen is to rerun the last block against the application again. Feel free to correct me if I've missed something.

I think the PR originally want the Tendermint core can have a way to roll back the case when the state is broken at H (appHash mismatch) and has a block stored at H+1. It's for sure we don't need rollback more than once if everything runs well.

Also see: https://github.com/cosmos/cosmos-sdk/pull/11361

**Comment by yihuang** [comment]
> It's not so much a concern as I fail to see how rolling back the last block could be useful the moment you restart, the node is just going to fetch that same block again. The only thing that needs to happen is to rerun the last block against the application again. Feel free to correct me if I've missed something.

the main use case right now is to avoid resync when the binary has unexpected breaking changes, so when some node crash for app hash mismatch, he can rollback, fix the binary and move on.

**Comment by yihuang** [comment]
don't stale.
there's unit test failure, @JayT106

**Comment by cmwaters** [comment]
> the main use case right now is to avoid resync when the binary has unexpected breaking changes, so when some node crash for app hash mismatch issue, he can rollback, fix the binary and move on.
and it's very useful for debugging such issues in devnet/testnet.

Yes that was the intention of the original rollback functionality which achieved this by removing the last "tendermint" state that it had saved. My question is what else do we gain from rolling back the block as well as the "tendermint" state. Rolling back shouldn't affect the content of the blocks themselves but rather just enable replaying it against the application binary

**Comment by JayT106** [comment]
> don't stale. there's unit test failure, @JayT106

rebased it on top of the `master`, the unit test failed was not related to the PR change.

**Comment by JayT106** [comment]
> > the main use case right now is to avoid resync when the binary has unexpected breaking changes, so when some node crash for app hash mismatch issue, he can rollback, fix the binary and move on.
> > and it's very useful for debugging such issues in devnet/testnet.
> 
> Yes that was the intention of the original rollback functionality which achieved this by removing the last "tendermint" state that it had saved. My question is what else do we gain from rolling back the block as well as the "tendermint" state. Rolling back shouldn't affect the content of the blocks themselves but rather just enable replaying it against the application binary

say if operators are having node crash issues related to some uncertain bugs in the Cosmos SDK or App. The content of the new block might be incorrect, also the app state might need to roll back for letting the node can continue to sync up to the network. Otherwise, the node will need to resync the network from the recent snapshot or from scratch which is painful for the node operators. Maybe there is no benefit from the Tendermint perspective, but it does very help in the Cosmos SDK and the App side.

**Comment by yihuang** [comment]
> > the main use case right now is to avoid resync when the binary has unexpected breaking changes, so when some node crash for app hash mismatch issue, he can rollback, fix the binary and move on.
> > and it's very useful for debugging such issues in devnet/testnet.
> 
> Yes that was the intention of the original rollback functionality which achieved this by removing the last "tendermint" state that it had saved. My question is what else do we gain from rolling back the block as well as the "tendermint" state. Rolling back shouldn't affect the content of the blocks themselves but rather just enable replaying it against the application binary

When there's a pending block (`block store height = state height + 1`), the current logic won't do anything, in this case, we must delete a block from block store first then rollback the state.

**Comment by cmwaters** [comment]
> When there's a pending block (block store height = state height + 1), the current logic won't do anything, in this case, we must delete a block from the block store first then rollback the state, this is the main motivation for rollback the block store.

if the block store height is greater than the state height that means that the node got the same app hash as the rest of the network (else it would have already panicked before persisting the new block) so there can't have been an app hash divergence (feel free to correct me if I'm wrong though)

**Comment by cmwaters** [comment]
> I think the block save happens before the panic, that's the case I have met before.

Does the panic come from an app hash mismatch or because of something during the execution of a block (i.e. `ApplyBlock`)

**Comment by yihuang** [comment]
> > I think the block save happens before the panic, that's the case I have met before.
> 
> Does the panic come from an app hash mismatch or because of something during the execution of a block (i.e. `ApplyBlock`)

https://github.com/tendermint/tendermint/blob/9dae97d845f1dffb134762e9b39e55d8009fe677/internal/state/validation.go#L54
I think it's this error, the `validateBlock` is called inside the `ApplyBlock` I think?

**Comment by cmwaters** [comment]
FYI opened this: https://github.com/tendermint/tendermint/pull/8493

**Comment by yihuang** [comment]
> FYI opened this: #8493

yeah, I guess this can fix the app hash mismatch case.

**Comment by cmwaters** [comment]
Is there anything from this PR we should still look to merge?

**Comment by JayT106** [comment]
> Is there anything from this PR we should still look to merge?

@cmwaters thanks for the fix. We think it should be good for now. The node should not fall into the case when syncing the pending block with a problem APP state.

**Pull request closed without merging**